### ISSUE TITLE
feat(optimizer): 2-phase aggregation

### ISSUE THREAD

**Issue by kwannoel** [singularity-data/risingwave#2638]
- For queries with few or no group keys, use 2-Phase aggregation.
- Otherwise use shuffle aggregation.

TODO: Update with implementation design.

cc @fuyufjh

**Comment by kwannoel** [comment]
Note that I'm taking some time to figure out how optimizer works, reading through testcases and so on. If this is an urgent feature, feel free to take it up / reassign it as I might not be done so soon :pray:

**Comment by skyzh** [comment]
We also have a lot of plan examples back to the time of legacy Java frontend... Anyone working on this issue may also take a look:

https://github.com/singularity-data/risingwave/blob/v0.1.7/legacy/planner/src/test/resources/com/risingwave/planner/planner/streaming/StreamAggPlannerDistributedModeTest.xml
https://github.com/singularity-data/risingwave/blob/v0.1.7/legacy/planner/src/test/resources/com/risingwave/planner/HashAggPlannerTest.xml

Basically, the local phase will contain the original agg functions (e.g., sum, count), while the global phase will have `sum` for `count`. e.g, select count(*) from table will have local phase `count`, and global phase to sum all counts to get the total row count.

**Comment by skyzh** [comment]
But the plan doesn't need to be exactly the same as Java frontend. IIRC, the `streaming_null_by_count` function has been removed, so simple agg's plan in the current frontend might differ from the legacy one.
